### PR TITLE
[training] Add deterministic replay for silent data corruption detection

### PR DESCRIPTION
**Pull request by xmfan**
Stacked PRs:
 * __->__#4324

--- --- ---

[training] Add deterministic replay for silent data corruption detection

## Summary

Add opt-in deterministic replay for detecting silent data corruption (SDC)
during training.

For selected optimizer steps, SDC replay:

1. snapshots the relevant trainer state;
2. executes a reference forward/backward unit;
3. restores the snapshot and executes the configured number of candidates;
4. compares the resulting signatures across all ranks; and
5. raises `SDCReplayMismatch` before the optimizer, scheduler, or checkpoint
   can observe corrupted state.

The signature covers:

- loss;
- local parameter gradients;
- registered module buffers;
- Python, CPU, and accelerator RNG state; and
- the trainer token counter.

Only the first gradient-accumulation unit of each checked optimizer step is
replayed. For pipeline parallelism, that unit includes the complete pipeline
schedule. The reference and intermediate candidates are discarded, while the
final matching candidate is committed.

## Configuration

The feature is disabled by default and can be enabled through the CLI:

```bash
--debug.deterministic
--sdc-replay.enabled
--sdc-replay.num-steps 1
--sdc-replay.num-replays 1
```

`num_steps` defaults to `1`; `-1` checks every locally attempted optimizer
step. `num_replays` defaults to `1` and counts candidates compared against the
reference.

Replay supports eager execution, `torch.compile`, CUDA graphs, GraphTrainer,
TorchFT, and the existing distributed execution modes. CUDA graphs currently
require `num_replays=1` because additional replays would need CUDA-graph-aware
restoration of graph-owned gradient and buffer storage.

Tensor comparisons use device-side `torch.hash_tensor` digests and perform a
single host synchronization on the matching path. The current XOR-based hash
is order-insensitive and may therefore have collisions.

## Testing

- `pytest -q tests/unit_tests/test_trainer.py tests/unit_tests/test_sdc_replay.py tests/unit_tests/test_invalid_loss.py tests/unit_tests/test_config_manager.py`
  - 69 passed
- `tests/integration_tests/features.py::sdc_replay_cudagraph`
  - Passed on 1 GPU
- `torchtitan/experiments/graph_trainer/tests/integration_tests.py::sdc_replay`
  - Passed on 1 GPU
- `tests/integration_tests/h100.py::deepseek_v3_fsdp+cp+tp+minimal_async_ep+sdc_replay`
  - Passed for 10 steps on 8 H100 GPUs with identical losses and gradient norms across ranks
- `tests/integration_tests/features.py::sdc_replay_mismatch`
  - Detected the injected rank-0 gradient mismatch on both ranks before the optimizer step
- `git diff --check`
- `pre-commit run --all-files`

### CODE REVIEW

**Comment by xmfan** [comment]
splitting this out into https://github.com/pytorch/torchtitan/pull/4333
